### ISSUE TITLE
To Do List/Priorities

### ISSUE THREAD

**Issue by tobeycarman** [ua-snap/dvm-dos-tem#194]
As of 7/15/2016, here is the list of priorities we came up with, with an attempt at listing the steps in order of priority:

- [ ] Attempt to implement "force closing" w/in C++ code. This finding a place w/in the C++ code, right after the integration finishes, to check that the cycles are closing and if not, force that to happen by choosing either the pool or fluxes as "truth".

- [ ] Finish implementing and testing the stage transitions so that we can run thru the transient stage and test the calibration.

- [ ] Calibrate with PEST. This is multi-faceted. Will require first getting the code running with PEST, then we will have to be able to run thru the transient stage to evaluate how well the model is doing matching the historic measurements.

- [ ] Finish implementing/refactoring the disturbance model (#119). First step will be re-basing the branch onto the current devel.

- [ ] Implement the NetCDF outputs, which will be needed for regional runs.

- [ ] Check that we are indeed able to run over a regional area. Historically this was a different "mode" from a single site run, so there was always a question about whether or not we could get things running in a "regional mode". After the refactor this shouldn't be too much of an issue.

**Comment by tobeycarman** [comment]
Made a bunch of progress with getting dvmdostem to run under PEST++ and SLURM on atlas, so now run-times are actually small enough that we can experiment with PEST++ in earnest!

As far as I can tell the next critical path item is stage transitions so that we can run out thru the transient stage and evaluate the quality of a calibration. 

The whole setup with pest++ on atlas and slurm is still a little cumbersome, but I tried to put a bunch of notes in the the `slurm-job.sh` script that is now in the `pestdemo/tussock_full` directory: https://github.com/tobeycarman/pestdemo/blob/tussock_full-take2/tussock_full/slurm-job.sh

**Comment by hgenet** [comment]
Great news Tobey. I know Ruth is working on the stage transition right 
now. Maybe she could give you an update on where she's at for 
coordination between you guys.

----
Hélène Genet, PhD
Institute of Arctic Biology
University of Alaska Fairbanks
902 Koyukuk Drive
Fairbanks, AK 99775 - USA
Phone: +1 907 699 4340
Email: hgenet@alaska.edu
Skype: helenegenet

**Comment by tobeycarman** [comment]
Yeah, it'd be great to get an update on the stage transition progress, as I will be looking into that next. It has been a while for me and I have forgotten what the issues are. If I recall we are pretty close to having it working and hopefully have just a few wrinkles to work out. When I figure out what specific issues I have, I will make a new github issue.

I copied the outputs from my first PEST++ run to `atlas` so you can all take a look at them:

```
/big_scratch/tem/initial-pest-outputs/master-00000/ 
```

It would be great if any/all of you could start familiarizing yourselves with the PEST output files so that we can know what to do next. The best file to start with would be the `tussock_full.rec` file which is the PEST "run record" file. It is the summary of the PEST run and is more or less human decipherable ;-).

**Comment by tobeycarman** [comment]
OK, I got the stage transitions working, and added some notes to the wiki page for ["Run Stages"] (https://github.com/ua-snap/dvm-dos-tem/wiki/Run-Stages).

Looks good @rarutter , as you has indicated a while ago, this is basically finished/working. It took me a while to figure out or remember that I had to create the restart file "templates" before it would successfully change from one stage to the next.

**Comment by tobeycarman** [comment]
Now I am finding myself wanting a way to view the outputs from subsequent stages on one time-serises, end to end, to be able to quickly make sure that each stage seems to be picking up where the last stage left off. Anyone have ideas for an easy way to do this?

**Comment by tobeycarman** [comment]
maybe enable the inter-stage pause, archive the calibration json files at the end of each stage, and add a bit of code to `calibration-viewer.py` so it can read from multiple archives?

**Comment by tobeycarman** [comment]
Well that wasn't too bad. I enabled the inter-stage pause. Then ran the model with this command:

    $ ./dvmdostem -l note -p 10 -m 20 -s 30 -t 25 -n 20 -c

Then I modified the `calibration-viewer.py` (in a really temporary manner; need to clean up my code before committing anything). Then at each inter-stage pause I ran a command to archive the calibration json outputs:

    $ tar -cjf pr-data.tar.gz /tmp/dvmdostem/
    ..... continue model, wait for equilibrium stage to finish...
    $ tar -cjf eq-data.tar.gz /tmp/dvmdostem/
    ..... continue model, wait for spinup stage to finish...
    $ tar -cjf sp-data.tar.gz /tmp/dvmdostem/
    ..... continue model, wait for transient stage to finish...
    $ tar -cjf tr-data.tar.gz /tmp/dvmdostem/
    ..... continue model, wait for scenario stage to finish...
    $ tar -cjf sc-data.tar.gz /tmp/dvmdostem/

And finally ran the modified `calibration-viewer.py` to get this plot:

    $ ./calibration/calibration-viewer.py --from-archive pr-data.tar.gz

<img width="1064" alt="screen shot 2016-08-17 at 4 53 27 pm" src="https://cloud.githubusercontent.com/assets/838735/17758368/2e4a70f2-649b-11e6-9c0f-25c60ba7eb3d.png">

Which looks pretty reasonable at first glance.
 - It looks like pr and eq stages are using a static climate
 - it looks like sp and tr are using the same climate. In the long run, this is likely not what we want, but due to some settings that have been in place for development, this is expected.
 - it looks like scenario is using a different climate, which is good

**Comment by rarutter** [comment]
As discussed, for now I implemented stage-specific JSON output storage. NetCDF will be an addition to that later.

Also, I verified that it will run for multiple cells (2) just with a change in the run mask. Still need to implement separate output for this, though.

**Comment by rarutter** [comment]
Item 4 has been addressed by the PRs listed in the thread for issue #119 

Item 5 is handled by PRs #257 and #265 

Item 6 has been resolved by the refactor.
